File: 259__3Sum_Smaller.py
class Solution(object):
    def threeSumSmaller(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        result = 0
        nums.sort()
        n = len(nums)
        for head in range(n - 2):
            mid = head + 1
            tail = n - 1
            if 3 * nums[head] >= target:
                break
            while mid < tail:
                s = nums[mid] + nums[head] + nums[tail]
                if s < target:
                    result += tail - mid
                    mid += 1
                else:
                    tail -= 1
        return result

File: Java/259__3Sum_Smaller.java
class Solution {
    public int threeSumSmaller(int[] nums, int target) {
        if (nums.length < 3) return 0;
        Arrays.sort(nums);
        int result = 0;
        for (int i = 0; i < nums.length - 2; i++) {
            if (nums[i] + nums[i + 1] + nums[i + 2] >= target) break;
            int j = i + 1, k = nums.length - 1;
            while (j < k) {
                if (nums[i] + nums[j] + nums[j + 1] >= target) break;
                while (j < k && nums[i] + nums[j] + nums[k] >= target) k--;
                result += k - j;
                j++;
            }
        }
        return result;
    }
}
